### PR TITLE
Handle collection redirects for filters/tests

### LINKED ISSUE #77192: moving ipaddr from netcommon to utils breaks the non-namespaced usage of nthhost (without the collection)
### Summary

https://github.com/ansible-collections/ansible.netcommon/pull/359 This change breaks the non-namespaced usage of ipaddr filter (without the collection). example ipwrap

### Issue Type

Bug Report

### Component Name

ansible

### Ansible Version

```console
$ ansible --version
code and can become unstable at any point.
ansible [core 2.13.0.dev0] (bugfix_redirection de11a1ce78) last updated 2022/03/03 13:38:18 (GMT +550)
  config file = /Users/amhatre/ansible-collections/playbooks/ansible.cfg
  configured module search path = ['/Users/amhatre/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/amhatre/dev-workspace/ansible/lib/ansible
  ansible collection location = /Users/amhatre/ansible-collections/collections
  executable location = /Users/amhatre/dev-workspace/ansible/bin/ansible
  python version = 3.8.5 (default, Jan  7 2021, 17:04:44) [Clang 12.0.0 (clang-1200.0.32.28)]
  jinja version = 3.0.3
  libyaml = True
```

### Configuration

```console
$ (ansible_from_source1) amhatre@ashwinis-MacBook-Pro playbooks % ansible-config dump --only-changed
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you
are modifying the Ansible engine, or trying out features under development. This is a rapidly changing source of
code and can become unstable at any point.
COLLECTIONS_PATHS(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = ['/Users/amhatre/ansible-collection
DEFAULT_HOST_LIST(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = ['/Users/amhatre/ansible-collection
HOST_KEY_CHECKING(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = False
INTERPRETER_PYTHON(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = /Users/amhatre/ansible_venvs/py3.8
PARAMIKO_LOOK_FOR_KEYS(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = False
PERSISTENT_COMMAND_TIMEOUT(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = 100
PERSISTENT_CONNECT_RETRY_TIMEOUT(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = 100
PERSISTENT_CONNECT_TIMEOUT(/Users/amhatre/ansible-collections/playbooks/ansible.cfg) = 100
(END)
```

### OS / Environment

mac os

### Steps to Reproduce

```yaml (paste below)
- name: Some configuration files require IPv6 addresses to be “wrapped” in square brackets ([ ]) This filter is used to wrap ipv6 address.
  hosts: localhost
  connection: ansible.netcommon.network_cli
  gather_facts: no
  tasks:
    - name: Input for IPVwrap plugin
      ansible.builtin.set_fact:
       value:
            - 192.24.2.1
            - host.fqdn
            - ::1
            - ''
            - 192.168.32.0/24
            - fe80::100/10
            - 42540766412265424405338506004571095040/64
            - True
    - debug:
        msg: "{{ value|ipwrap }}"
```

### Expected Results

(py3.6.10) amhatre@ashwinis-MacBook-Pro playbooks % ansible-playbook test_ipwrap.yaml

PLAY [Some configuration files require IPv6 addresses to be “wrapped” in square brackets ([ ]) This filter is used to wrap ipv6 address.] ***

TASK [Input for IPVwrap plugin] *********************************************************************************
ok: [localhost]

TASK [debug] ****************************************************************************************************
[WARNING]: The value '' is not a valid IP address or network, passing this value to ipaddr filter might result
in breaking change in future.
[WARNING]: The value 'True' is not a valid IP address or network, passing this value to ipaddr filter might
result in breaking change in future.
ok: [localhost] => {
    "msg": [
        "192.24.2.1",
        "host.fqdn",
        "[::1]",
        "",
        "192.168.32.0/24",
        "[fe80::100]/10",
        "[2001:db8:32c:faad::]/64",
        true
    ]
}

PLAY RECAP ******************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

(py3.6.10) amhatre@ashwinis-MacBook-Pro playbooks % 

### Actual Results

```console
(py3.6.10) amhatre@ashwinis-MacBook-Pro playbooks % ansible-playbook test_ipwrap.yaml

PLAY [Some configuration files require IPv6 addresses to be “wrapped” in square brackets ([ ]) This filter is used to wrap ipv6 address.] ************************

TASK [Input for IPVwrap plugin] **********************************************************************************************************************************
ok: [localhost]

TASK [debug] *****************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "template error while templating string: No filter named 'ipwrap'.. String: {{ value|ipwrap }}"}

PLAY RECAP *******************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

(py3.6.10) amhatre@ashwinis-MacBook-Pro playbooks %
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Based on https://github.com/ansible/ansible/issues/77192#issuecomment-1058159370

Fixes #77192
Closes #77194

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/template/__init__.py

### CODE REVIEW

**Comment by s-hertel** [comment]
Most of this diff is just indentation. You can see the important parts more easily by excluding whitespace changes: https://github.com/ansible/ansible/pull/77210/files?w=1

File: lib/ansible/template/__init__.py [hunk 1]
```diff
@@ -469,56 +470,61 @@ def __getitem__(self, key):
...
                     return func
 
-                # didn't find it in the pre-built Jinja env, assume it's a former builtin and follow the normal routing path
-                leaf_key = key
-                key = 'ansible.builtin.' + key
-            else:
-                leaf_key = key.split('.')[-1]
+            key, leaf_key = get_fqcr_and_name(key)
+
+            for dummy in range(30):
```
Comment (sivel) [inline lib/ansible/template/__init__.py]: Curious about your thoughts on this.  I think generally I'm fine with it as is, but should we not care about depth, and instead track redirects to ensure they aren't duplicates?

I tried to look at how this is done in the plugin loader, but it's just not implemented the same here and there.

Comment (s-hertel) [reply lib/ansible/template/__init__.py]: That seems like a good idea. I'll look closer at how the plugin loader does this and tweak this.

Comment (s-hertel) [reply lib/ansible/template/__init__.py]: Updated this to loop indefinitely until the same fqcr is encountered twice. This is what the plugin loader does too - it checks if the redirect is already present in the list of redirects accumulated so far.

**Review by sivel — APPROVED** [review]

**Comment by ashwini-mhatre** [comment]
@s-hertel I tested your code. its working as expected

**Comment by ashwini-mhatre** [comment]
@sivel when this patch will be available in ansible code? is there any milestone? CC: @trishnaguha

**Comment by s-hertel** [comment]
@ashwini-mhatre Thank you for checking! I'm going to fix up a couple variable names in the tests and then I'll merge this today if there are no objections and then open backports for 2.11 and 2.12. The next release date is March 28th as per https://github.com/ansible/ansible/pull/77194#issuecomment-1059212781.

**Comment by s-hertel** [comment]
Backports created.

**Pull request merged**